src/quadratic_expression.cpp

```diff
@@ -32,8 +32,8 @@ Rcpp::List quadratic_expression(Eigen::Map<Eigen::VectorXd> X,
     // Auto differential.
     auto f = ad::autodiff(expr, seed.array());
 
-    Rcpp::List res = Rcpp::List::create(Rcpp::Named("value") = f,
-                                        Rcpp::Named("gradient") = x.get_adj());
+    Rcpp::List res = Rcpp::List::create(Rcpp::Named("value") = f(0,0),
+                                        Rcpp::Named("gradient") = x.get_adj().col(0));
```

Comment:
This can be avoided if we do
```cpp
Eigen::VectorXd x_adj(2);
x_adj.setZero();
ad::VarView<double, ad::vec> x(X.data(), x_adj.data(), 2);
```

Reply:
Can you maybe just take the branch as is and stick all the proposed changes in?  And then I glance at it?

Reply:
Or else spell it out more clearly and I put it in / adjust tests as needed too.